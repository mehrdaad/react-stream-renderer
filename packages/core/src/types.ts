export type TextTransform = 'none' | 'capitalize' | 'uppercase' | 'lowercase';

export type Style = {
  color?: string;
  bgColor?: string;
  modifiers?: string[];
  textTransform?: TextTransform;
};

export type RectData = {
  absolute: boolean;
  x: number;
  y: number;
  z: number;
  relativeX: number;
  relativeY: number;
  width: number;
  height: number;
};

export type OnLayoutHook = (layout: RectData) => void;

export type MouseEvent = {
  x: number;
  y: number;
  code: number;
  shift: boolean;
  alt: boolean;
  ctrl: boolean;
};

export type OnClickHook = (event: MouseEvent) => boolean | void;

export type JsonText = {
  node: 'Text';
  body: string;
  width: number;
  height: number;
  x: number;
  y: number;
  style?: Style;
};

export type JsonParagraph = {
  node: 'Paragraph';
  children: Array<JsonParagraph | JsonText>;
  width: number;
  height: number;
  x: number;
  y: number;
  style?: Style;
};

export type JsonView = {
  node: 'View';
  children: Array<JsonView | JsonText>;
  width: number;
  height: number;
  x: number;
  y: number;
  style?: Style;
};
