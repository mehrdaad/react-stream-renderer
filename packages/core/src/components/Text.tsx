import React from 'react';
import Paragraph from '../host/nodes/Paragraph';
import { capitalize } from '../utils';
import {
  SingleOrMulti,
  TextTransform,
  OnLayoutHook,
  MouseEventHandler,
} from '../types';

export type Props = {
  children: SingleOrMulti<number | string | JSX.Element>;
  color?: string;
  bgColor?: string;
  textTransform?: TextTransform;
  dim?: boolean;
  underline?: boolean;
  lineThrough?: boolean;
  bold?: boolean;
  italic?: boolean;
  onLayout?: OnLayoutHook;
  onClick?: MouseEventHandler;
  onWheel?: MouseEventHandler;
};

export default function View(props: Props) {
  return (
    // @ts-ignore
    <Paragraph.TAG
      onLayout={props.onLayout}
      onClick={props.onClick}
      onWheel={props.onWheel}
      style={{
        color: props.color,
        bgColor: props.bgColor,
        modifiers: reduceModifiers(props),
      }}
    >
      {React.Children.map(props.children, child => {
        if (React.isValidElement(child)) {
          return React.cloneElement(child, {
            textTransform: props.textTransform,
          } as any);
        }

        switch (props.textTransform) {
          case 'capitalize': {
            return capitalize(child.toString());
          }
          case 'lowercase': {
            return child.toString().toLowerCase();
          }
          case 'uppercase': {
            return child.toString().toUpperCase();
          }
          case 'none':
          default:
            return child;
        }
      })}
    </Paragraph.TAG>
  );
}

function reduceModifiers(props: Props) {
  const modifiers = [];
  if (props.bold) {
    modifiers.push('bold');
  }
  if (props.italic) {
    modifiers.push('italic');
  }
  if (props.dim) {
    modifiers.push('dim');
  }
  if (props.underline) {
    modifiers.push('underline');
  }
  if (props.lineThrough) {
    modifiers.push('strikethrough');
  }
  return modifiers.length ? modifiers : undefined;
}
