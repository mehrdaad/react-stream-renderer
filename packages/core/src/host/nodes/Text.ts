import assert from 'assert';
import stripAnsi from 'strip-ansi';
import Base from './Base';

export type TextStyle = {
  color?: string;
  bgColor?: string;
  modifiers?: string[];
};

export default class Text extends Base<never> {
  static TAG = 'TEXT_NODE';

  private body: string = '';

  setBody(body: string) {
    assert(!body.includes('\n'), 'Text body cannot have new line characters');
    this.body = body;
    this.layoutNode.setStyle({
      width: stripAnsi(this.body).length,
      height: 1,
    });
  }

  getBody(): string {
    return this.body;
  }
}
