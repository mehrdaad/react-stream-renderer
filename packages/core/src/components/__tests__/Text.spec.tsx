import React from 'react';
import Text from '../Text';
import renderToJson from '../../renderToJson';
import { JsonText } from '../../types';

describe('Text component', () => {
  it('should transform children', () => {
    expect(
      (renderToJson(<Text textTransform="uppercase">hello</Text>).snapshot
        .children[0] as JsonText).body
    ).toEqual('HELLO');
    expect(
      (renderToJson(<Text textTransform="lowercase">HELLO</Text>).snapshot
        .children[0] as JsonText).body
    ).toEqual('hello');
    expect(
      (renderToJson(<Text textTransform="capitalize">hello world</Text>)
        .snapshot.children[0] as JsonText).body
    ).toEqual('Hello World');
    expect(
      (renderToJson(<Text textTransform="none">hello</Text>).snapshot
        .children[0] as JsonText).body
    ).toEqual('hello');
    expect(
      (renderToJson(
        <Text textTransform="uppercase">{['hello ', 'world']}</Text>
      ).snapshot.children[0] as JsonText).body
    ).toEqual('HELLO WORLD');
  });

  it('should add modifiers', () => {
    expect(
      renderToJson(<Text>hello</Text>).snapshot.children[0].style.modifiers
    ).toBeUndefined();
    expect(
      renderToJson(<Text bold>hello</Text>).snapshot.children[0].style.modifiers
    ).toEqual(['bold']);
    expect(
      renderToJson(<Text dim>hello</Text>).snapshot.children[0].style.modifiers
    ).toEqual(['dim']);
    expect(
      renderToJson(<Text underline>hello</Text>).snapshot.children[0].style
        .modifiers
    ).toEqual(['underline']);
    expect(
      renderToJson(<Text lineThrough>hello</Text>).snapshot.children[0].style
        .modifiers
    ).toEqual(['strikethrough']);
    expect(
      renderToJson(<Text italic>hello</Text>).snapshot.children[0].style
        .modifiers
    ).toEqual(['italic']);

    expect(
      renderToJson(
        <Text bold italic>
          hello
        </Text>
      ).snapshot.children[0].style.modifiers
    ).toEqual(['bold', 'italic']);
  });

  it('should set color and bgColor', () => {
    expect(
      renderToJson(
        <Text color="red" bgColor="blue">
          hello
        </Text>
      ).snapshot.children[0].style
    ).toEqual({ color: 'red', bgColor: 'blue', modifiers: undefined });
  });
});
