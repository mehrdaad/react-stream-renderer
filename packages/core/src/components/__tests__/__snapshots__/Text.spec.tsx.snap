// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text component should transform children 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "body": "hello",
          "height": 1,
          "node": "Text",
          "style": undefined,
          "width": 5,
          "x": 0,
          "y": 0,
        },
      ],
      "height": 1,
      "node": "Paragraph",
      "style": Object {
        "bgColor": undefined,
        "color": undefined,
        "modifiers": undefined,
        "textTransform": "uppercase",
      },
      "width": 5,
      "x": 0,
      "y": 0,
    },
  ],
  "height": 1,
  "node": "View",
  "style": undefined,
  "width": 5,
  "x": 0,
  "y": 0,
}
`;

exports[`Text component should transform children 2`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "body": "HELLO",
          "height": 1,
          "node": "Text",
          "style": undefined,
          "width": 5,
          "x": 0,
          "y": 0,
        },
      ],
      "height": 1,
      "node": "Paragraph",
      "style": Object {
        "bgColor": undefined,
        "color": undefined,
        "modifiers": undefined,
        "textTransform": "lowercase",
      },
      "width": 5,
      "x": 0,
      "y": 0,
    },
  ],
  "height": 1,
  "node": "View",
  "style": undefined,
  "width": 5,
  "x": 0,
  "y": 0,
}
`;

exports[`Text component should transform children 3`] = `
Object {
  "snapshot": Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "body": "hello world",
            "height": 1,
            "node": "Text",
            "style": undefined,
            "width": 11,
            "x": 0,
            "y": 0,
          },
        ],
        "height": 1,
        "node": "Paragraph",
        "style": Object {
          "bgColor": undefined,
          "color": undefined,
          "modifiers": undefined,
          "textTransform": "capitalize",
        },
        "width": 11,
        "x": 0,
        "y": 0,
      },
    ],
    "height": 1,
    "node": "View",
    "style": undefined,
    "width": 11,
    "x": 0,
    "y": 0,
  },
  Symbol(Symbol.iterator): [Function],
}
`;

exports[`Text component should transform children 4`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "body": "hello",
          "height": 1,
          "node": "Text",
          "style": undefined,
          "width": 5,
          "x": 0,
          "y": 0,
        },
      ],
      "height": 1,
      "node": "Paragraph",
      "style": Object {
        "bgColor": undefined,
        "color": undefined,
        "modifiers": undefined,
        "textTransform": "none",
      },
      "width": 5,
      "x": 0,
      "y": 0,
    },
  ],
  "height": 1,
  "node": "View",
  "style": undefined,
  "width": 5,
  "x": 0,
  "y": 0,
}
`;

exports[`Text component should transform children 5`] = `
Object {
  "snapshot": Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "body": "hello ",
            "height": 1,
            "node": "Text",
            "style": undefined,
            "width": 6,
            "x": 0,
            "y": 0,
          },
          Object {
            "body": "world",
            "height": 1,
            "node": "Text",
            "style": undefined,
            "width": 5,
            "x": 6,
            "y": 0,
          },
        ],
        "height": 1,
        "node": "Paragraph",
        "style": Object {
          "bgColor": undefined,
          "color": undefined,
          "modifiers": undefined,
          "textTransform": "uppercase",
        },
        "width": 11,
        "x": 0,
        "y": 0,
      },
    ],
    "height": 1,
    "node": "View",
    "style": undefined,
    "width": 11,
    "x": 0,
    "y": 0,
  },
  Symbol(Symbol.iterator): [Function],
}
`;
